test-projects/gtm-test/cypress/integration/iubenda-gtm-client-side-nav.spec.js

```diff
@@ -20,22 +19,21 @@ context('Iubenda GTM client-side navigation', () => {
       win.dataLayer = [];
     });
 
-    cy.get(".page-2-link").click();
+    cy.contains('Go to page 2').click();
 
     cy.window().its('dataLayer').should('contain', { event: "iubenda_consent_given" });
   })
 
-  it('The Iubenda cookie footer should NOT push the event to the GTM data layer if consent has NOT been given and the user navigates to a different page (onRouteUpdate)', () => {
+  it('The Iubenda cookie footer should not push the event to the GTM data layer if consent has not been given and the user navigates to a different page', () => {
     cy.visit('http://localhost:9000');
     cy.get("#iubenda-cs-banner").should("exist");
     cy.get(".iubenda-cs-close-btn").should("exist");
-    cy.get(".page-2-link").should("exist");
 
     cy.window().then(win => {
       win.dataLayer = [];
     });
 
-    cy.get(".page-2-link").click();
+    cy.contains('Go to page 2').click();
 
     cy.window().its('dataLayer').should('not.contain', { event: "iubenda_consent_given" });
```

Comment:
I thought about that this night, tell me what do you think: generally speaking, this check is weak, what happens if the plugin pushes something in the array a millisecond after the check? The problem is not with your test, obviously, it's in the nature of testing something that does not happen 😉

How could we fix it? My idea is:
- the first test (the "positive" test, not the "negative" test we are talking about) should check that
  - if you visit the first page and you give the consent, the array contains the expected object
  - then, if you navigate to the second page, the array contains two times the expected object
- while the "negative" test should check that
  - if you visit the first page and you don't give the consent...
  - then, if you visit the second page and don't you give the consent...
  - then, if you visit the first page (or a third one, it doesn't matter) and you give the consent, the array contains the expected object
  - then, if you navigate to the second page (or a fourth one, it doesn't matter), the array contains two times the expected object

This way we would be completely sure that the first and the second navigation don't pollute the array! What do you think?

Reply:
You're right, there is a potential for a race condition here.

I agree that we can check for duplication on the positive test, however, I don't think the "negative" test should involve giving consent at any point - it seems to defeat the purpose of the negative test. By adding the third visit, I don't think it fixes the issue of the race condition either.

Perhaps we could add a third/fourth visit **without** giving consent, to give the event handler multiple chances to fire? We could also add a `cy.wait(1000)` but I would rather avoid this approach as it doesn't really solve the problem either.

What do you think?

Reply:
You're right. To have the `cy.wait` advantages without leaving the Cypress garden, the second test could:
- visit the first page
- go to the second page
- the test expects to fail ([here](https://github.com/NoriSte/cypress-wait-until/blob/master/cypress/integration/plugin.spec.js#L202) you can find a real world example)
```js
 cy.once('fail', (err) => {
    expect(err.message).to.be.equal('<PUT THE CYPRESS ERROR HERE>')
  })
```
- assert that the array should contain the expected object
- Cypress retries up to 5 seconds for the array check, then it fails
- thanks to `cy.once('fail' // ...` our test will pass

Doing this: we mitigate the race condition (if the array is empty even after 5 seconds we could be enough sure that Iubenda won't push the object at all) and we stay in what Cypress offers. Do you agree or do you prefer going with your solution? 😊

Reply:
I like this! Will give that a try. Thanks